### PR TITLE
Fix multiple api calls on route change

### LINKED ISSUE #2482: Repeated requests to the server/api endpoint.
**Describe the bug**
Repeated requests are made to the `server/api` endpoint when loading a page. This is unlikely to have a good justification as the data provided on the root endpoint is practically static. It gives basic information about the repository and lists its endpoints (endpoints shortened for brevity.

```json
{
  "dspaceUI": "https://sandbox.dspace.org",
  "dspaceName": "DSpace Sandbox",
  "dspaceServer": "https://sandbox.dspace.org/server",
  "dspaceVersion": "DSpace 8.0-SNAPSHOT",
  "type": "root",
  "_links": {
    "actuator": {
      "href": "https://sandbox.dspace.org/server/actuator"
    },
    "authn": {
      "href": "https://sandbox.dspace.org/server/api/authn"
    },
  }
}
```

The following table lists the number of requests different pages make to that endpoint (these are links to the public server, some of which may at any time become unavailable).

| Page      | Number of requests to  `server/api` |
| ----------- | ----------- |
| https://sandbox.dspace.org/home      | 2       |
| https://sandbox.dspace.org/community-list   | 2        |
| https://sandbox.dspace.org/communities/8b632938-77c2-487c-81f0-e804f63e68e6   | 3       |
| https://sandbox.dspace.org/collections/ece42bb4-5ec0-4760-b727-30f9f593acb8   | 3       |
| https://sandbox.dspace.org/items/004286c4-60b8-44d7-94f7-a1658351e9a4   | 3        |
| https://sandbox.dspace.org/browse/dateissued   | 3        |
| https://sandbox.dspace.org/browse/title  | 3       |
| https://sandbox.dspace.org/entities/project/3329a278-2719-4bae-a5d5-d2039afc2fe2   | 3        |
| https://sandbox.dspace.org/statistics/items/3329a278-2719-4bae-a5d5-d2039afc2fe2   | 2        |
| https://sandbox.dspace.org/info/privacy  | 2        |

**To Reproduce**
1. Open any of the above URLs, with the network tab of your browser open
2. Sort requests by name and count the number of requests to `/server/api`

**Expected behavior**
Each page should presumably only need to request the `/server/api` endpoint once.

### PR DESCRIPTION
**Pull request by artlowel**
## References
* Fixes #2482 

## Description
The issue was caused by the fact that the guard made a new request to the api every time something subscribed to the result of its canActivateChild method. That happened more then once per route change

The solution was to use the cached /api response, and invalidate that cache on every NavigationStart event. That way, the first subscriber triggers the re-request, and all the others use the cached version.

## Instructions for Reviewers
Open the network tab in the browser's dev tools, and check that there's one a single /api request every time you change the route

Also verify that if the backend becomes unavailable, that you're redirected to the 500 error page the next time you change the route

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [ESLint](https://eslint.org/) validation using `yarn lint`
- [x] My PR doesn't introduce circular dependencies (verified via `yarn check-circ-deps`)
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

**Comment by alanorth** [comment]
Thanks @artlowel! I tested the patch by opening the network tab of my browser dev tools and clearing it before navigating to another route. Before applying I would see two requests to `/api`. After applying I only see one request to `/api`.

Also, as suggested, I stopped Tomcat and then the next time I tried to navigate to a new route I received the HTTP 500 error page.

So on functionality this looks good to me. I will let someone else who is better with Angular review the code.

**Comment by artlowel** [comment]
@alanorth I can reproduce your issue only when the backend is down. It makes sense that you can't log in when the backend is down. After it is back up, logging in works again for me.

I did however discover an bug where the login dropdown goes blank if you enter the wrong account details. But I can reproduce that on both https://demo.dspace.org/ and on https://sandbox.dspace.org/, so it's not related to this PR. I created an issue for it: https://github.com/DSpace/dspace-angular/issues/2515

**Comment by alanorth** [comment]
@artlowel I just tried again, and my backend is definitely up because I can browse everywhere in the repository, but my login dropdown is white. Shift refresh, nothing. See quick demo:


https://github.com/DSpace/dspace-angular/assets/191754/2da6a327-d063-422a-a4f8-bb09c89d36af


Another thing I noticed is that some browse pages actually make THREE calls to `/api`! Check "Browse by Issue Date" for example. With this patch they only make one. So that's a win.

**Comment by alanorth** [comment]
@artlowel if I revert the patch the white login dropdown issue goes away. Also note, in my video above, that I have not attempted login at all, so I'm not sure that bug is related to a failed login.

**Comment by artlowel** [comment]
@alanorth then that must mean that https://github.com/DSpace/dspace-angular/issues/2515 also happens for other errors, besides a failed login. 

Can you check your console when the blank dropdown issue happens. Is `undefined doesn't contain the link authn` still the error shown? The reason I ask is because that error tends to mean that `/api` can't be retrieved (as `authn` is a link on `/api`) and in your video we see that it _can_ be retrieved

**Comment by alanorth** [comment]
@artlowel yes, it is reliable. Right after applying the patch, Angular rebuilds, the page refreshed, and I get errors in the console. Three messages about missing links (one for authn, two for properties):

2023-09-25T15:39:34,452822531+03:00-fs8

**Comment by artlowel** [comment]
@alanorth I'm able to reproduce your issue now. It happens only in dev mode. I'll take a look

**Comment by J4bbi** [comment]
Just to quickly note that I get the same errors. Running in development mode with sandbox as API.

I don't know if it's an issue but the `CanActivateChild` interface used in [server-check.guard.ts](https://github.com/DSpace/dspace-angular/blob/e92b58eb9e7dc64fe0cf98028b59c9d175de3bbe/src/app/core/server-check/server-check.guard.ts#L17)  is deprecated in favour of "[functional injectable guards](https://angular.io/api/router/CanActivateChild)". Is this a concern? (I have no idea)

I put in some `console.traces()` and seems to me that a lot of services and functionality want to know how `server/ap`i is doing.

image

**Comment by artlowel** [comment]
@J4bbi 
>I don't know if it's an issue but the CanActivateChild interface used in [server-check.guard.ts](https://github.com/DSpace/dspace-angular/blob/e92b58eb9e7dc64fe0cf98028b59c9d175de3bbe/src/app/core/server-check/server-check.guard.ts#L17) is deprecated in favour of "[functional injectable guards](https://angular.io/api/router/CanActivateChild)". Is this a concern? (I have no idea)

It looks like [CanActivate itself](https://angular.io/api/router/CanActivate) is also deprecated, which means that eventually we'll need to update all our guards. I think this is something that is best done all at once, e.g. in the context of the next angular upgrade

**Comment by alanorth** [comment]
Hey @artlowel. I tested again with the updated patch and I am able to log in. I also confirm that the calls to `/api` are still reduced to one as expected. I think I'm +1, but can wait for @J4bbi to comment as well.

File: src/app/core/data/root-data.service.ts [hunk 1]
```diff
@@ -25,21 +24,21 @@ export class RootDataService extends BaseDataService<Root> {
     protected rdbService: RemoteDataBuildService,
     protected objectCache: ObjectCacheService,
     protected halService: HALEndpointService,
-    protected restService: DspaceRestService,
   ) {
-    super('', requestService, rdbService, objectCache, halService, 6 * 60 * 60 * 1000);
+    super('', requestService, rdbService, objectCache, halService, 60 * 1000);
```
Comment (J4bbi) [inline src/app/core/data/root-data.service.ts:28]: Do I understand correctly that the cached response now invalidates in 1 minute instead of 6 hours?

**Review by J4bbi — APPROVED** [review]
Just one question. Everything works as it should

File: src/app/core/data/root-data.service.ts (hunk 1 above)
Comment (artlowel) [reply src/app/core/data/root-data.service.ts:28]: Good catch! 

You're right: when I started looking into this issue, I was thinking about simply lowering the validity of that root endpoint request, but then I figured we'd still need something to do the job the guard is doing now (redirect when that request fails) so I ended up fixing the guard instead. 

I'll roll back this change to ensure it doesn't break anything unexpectedly

**Comment by alanorth** [comment]
Merging this as it has been tested by myself and @J4bbi, with our review feedback addressed.

I think if we want to add a comment about the six hours you can add that later and we can merge it without testing all that.

**Comment by artlowel** [comment]
@J4bbi 
>@artlowel , great. I just wonder, because that number seems somewhat arbitrary, if it's possible to leave a comment to explain why it's 6 hours and not something else, or if there is something dependent on it being at least a specific value?

It doesn't have to be any specific value no. The original idea was that it should basically never expire because all the /api endpoint was used for was to retrieve the links to all other endpoints, and those only change if major new features are added to dspace. So 6 hours was chosen as just "a long time".

With the introduction of the ServerCheckGuard, that this PR makes some tweaks to, that endpoint is used to check that the server is still running as well. So in practice those 6 hours will never be reached, because the timer will restart every time you switch routes. The reason I still rolled back the change though, is because I don't want to introduce issues where the validity expires while you're on a page and not doing anything. In theory, if everything else is properly designed to use dataservices and work with the cache that shouldn't be a problem, it would just cause /api to be re-requested automatically when that time is up. But I'd rather not risk introducing new edge cases that haven't been tested for

**Comment by alanorth** [comment]
@artlowel the automatic port to `main` failed. Would you be able to port it manually and make a PR? Assuming all tests pass we can merge it ASAP. Thanks!
